### ISSUE TITLE
Feature Request: time codes display as frame numbers

### ISSUE THREAD

**Issue by quentinjcm** [mpv-player/mpv#6568]
It could be useful to have a toggle for displaying time codes as frame numbers in the viewer. I couldn't see options for this in the reference manual.

**Comment by lachs0r** [comment]
Most containers don’t have a concept of numbered frames, only timestamps. Which also implies variable framerates, making this technically impossible to implement reliably. If you want a guesstimate (which won’t be guaranteed to be accurate even for files with constant frame duration), write a user script.

**Comment by qmega** [comment]
If you're fine with estimates, though, you might not need a user script; mpv already has estimated frame properties: `--osd-msg3='${estimated-frame-number} / ${estimated-frame-count}'`

https://mpv.io/manual/master/#command-interface-estimated-frame-count

**Comment by wabiloo** [comment]
This method doesn't work when using multiple videos and stacking them (with `--lavfi-complex` and `vstack/hstack`)

**Comment by NicTanghe** [comment]
Hello everyone.
does this mean that the frames displayed are always wrong with resolve ?

If they are displaying the accurate frame numbers reliably there should be a way to implement it.
also if you say most containers dont have this information that means some do and for those it should have the option.

i think mov has it, wich is basically the one that most requires it. (or is it only quicktime and is quicktime also using an inacurate numbers.

**Comment by Bellavene** [comment]
> If you're fine with estimates, though, you might not need a user script; mpv already has estimated frame properties: `--osd-msg3='${estimated-frame-number} / ${estimated-frame-count}'`

Can someone write a lua script which converts this to usual timecode based on fps in video? I don't think that this is impossible.

Also, why mpv can not read the timecode baked into files? This is so strange for me.

**Comment by llyyr** [comment]
`--script-opts-add=stats-show_frame_info=yes` then press `i` to see it.

> Also, why mpv can not read the timecode baked into files? This is so strange for me.

This is answered in the first response.

**Comment by Bellavene** [comment]
> `--script-opts-add=stats-show_frame_info=yes` then press `i` to see it.
> 
> > Also, why mpv can not read the timecode baked into files? This is so strange for me.
> 
> This is answered in the first response.

I meant not that, I meant [SMPTE timecode](https://en.wikipedia.org/wiki/SMPTE_timecode) in OSD

**Comment by llyyr** [comment]
> I meant not that, I meant [SMPTE timecode](https://en.wikipedia.org/wiki/SMPTE_timecode) in OSD and as variable ${current-timecode}

Click on the timestamp on the left on the OSC...

**Comment by llyyr** [comment]
I misread the wiki, it seems it needs to be in the format HH:MM:SS:Frame

so `--osd-msg1="SMPTE: ${time-pos/full}:${estimated-frame-number}"`

Remove the `/full` if you don't care about microseconds. Remember that the frame number is only an estimate and might be wrong.

**Comment by Bellavene** [comment]
Thank you. I it is ok, if it would estimate the timecode.
But this is not SMPTE, this just adds the overall frame count number at the end. While timecodes show frames within current second.

**Comment by Bellavene** [comment]
In fact, you could just translate milliseconds according to frames.
In case of 25 fps:
000 = 1
040 = 2
080 = 3
etc.

**Comment by llyyr** [comment]
```lua
ov = mp.create_osd_overlay("ass-events")
mp.observe_property("time-pos", "native", function(name, time_pos)
    if time_pos == nil then return end
    local s = math.floor(time_pos)
    ov.data = string.format("%02d:%02d:%02d:%02d", math.floor(s/3600), math.floor((s%3600)/60), s%60, math.floor((time_pos-s)*mp.get_property_native("container-fps")))
    ov:update()
end)
```

> I still can't get it, why it is not baked in.

Nobody sent a patch, you can be the one

**Comment by Bellavene** [comment]
I would gladly, but I make videos and have not much time to be a coder also. So I would rather pay you to implement it right.

**Comment by llyyr** [comment]
https://github.com/mpv-player/mpv/pull/14876

**Comment by NicTanghe** [comment]
Big thanks to everyone for doing implementing it so quietly after my last response.
It seems like the frameumber starts 0 instead of 1 somewhere maybe.

It is stil unclear to my where the inaccuracy on the frame numbers come from.
Are you stating that the recorded frame-numbers are incorrect ?

Or that in some meant for end use delivery format is simply unrealyable because of the compression ?

**Comment by llyyr** [comment]
> if fps is 24, instead of frame number 23 it draws 00. If fps is 25, it doesn't draw 09,11,13,15,17,19,21 frame numbers

It was a rounding issue, should be fixed.

> It is stil unclear to my where the inaccuracy on the frame numbers come from.

Should a 24/1.0001 frame rate file have a 24th frame? Should it not and why? What about files with variable frame rate? What about situations where the user is frame-back-stepping and mpv can't seek exactly to the previous frame? frame-back-stepping is a hack anyway and is generally impossible, it is for a reason VLC doesn't implement it for example.

I knew this property is impossible to get accurate with modern container formats, but I still tried and only seemed to prove myself right.

>  does this mean that the frames displayed are always wrong with resolve ?

I'm sure there are creative things you can do to have accurate frame timecode when you're _creating_ the video, but when it comes to deducing it from a video file, any player that claims to support it is doing guesswork that shouldn't be used for anything professional.

**Comment by Bellavene** [comment]
> Big thanks to everyone for patching it so quietly after my last response. It seems like the frameumber starts 0 instead of 1 somewhere maybe.

Frame numbers always start with 0

**Comment by llyyr** [comment]
> But if you look at milliseconds

milliseconds is not nearly accurate enough when it comes to determining frame duration, especially for matroska formats where nanoseconds are rounded on 1ms. Some muxers might go off by 1ms more, then correct it by an equal rounding error at a later frame in the opposite direction. mpv tries to "unround" these values as best as possible, and it's generally good enough for playback, but not for something like displaying a timecode with the current frame number perfectly accurately.

**Comment by NicTanghe** [comment]
> > Big thanks to everyone for patching it so quietly after my last response. It seems like the frameumber starts 0 instead of 1 somewhere maybe.
> 
> Frame numbers always start with 0
> 
> > I'm sure there are creative things you can do to have accurate frame timecode when you're creating the video, but when it comes to deducing it from a video file, any player that claims to support it is doing guesswork that shouldn't be used for anything professional.
> 
> @llyyr But if you look at milliseconds, they have this pattern 000, and they always follow the same amount on specific fps. So there is not much guessing, and it should be reliable enough. Anyway, Thank You again! Our hero.

frame numbers definately don`t always start at 0 in everyplace everywhere.
starting things from 0 is a programmer thing to do .

But if you tell me you know your shit and every video delivery-standard is encoded to start frame-numbers from 0 in video format i will accept that as truth.

**Comment by Bellavene** [comment]
> But if you tell me you know your shit and every video delivery-standard is encoded to start frame-numbers from 0 in video format i will accept that as truth.

Sounds a bit rude. Anyway, yeah, do you see any software which start it with 1, tell me. Thank you.

**Comment by NicTanghe** [comment]
maya.
every software that has shots that don't start at 0 although if the main
timeline still starts at 0 that doesn't really count.

I didn`t mean to insult your honour. knowing your shit is a positive trait.



Op wo 18 sep 2024 om 16:15 schreef Bellavene ***@***.***>:

**Comment by Bellavene** [comment]
Maya is a 3d software, not a video editor. While it starts at one, it just counts all frames without time, that's why. Ok I will rephrase, do you know any video software with a timeline which starts with 1?

**Comment by kasper93** [comment]
After some deliberation, we merged this https://github.com/mpv-player/mpv/commit/d33bcc51a75d32eaf76ec6b95fe758ab3c020b6b you can use three new properties

```
video-frame-info/gop-timecode
        String with the GOP timecode encoded in the frame.
video-frame-info/smpte-timecode
        String with the SMPTE timecode encoded in the frame.
video-frame-info/estimated-smpte-timecode
        Estimated timecode based on the current playback position and frame count.
```

**Comment by Bellavene** [comment]
Great! Thank You. One more thing.
Can we change frame `;` character to `:` ?  Would be much a help if you need to copy the timecode from mpv and paste it directly into editing software or EDL list if needed.

And will it take too much effort if mpv could read timecode track directly from the video file (if available)?

**Comment by llyyr** [comment]
> And will it take too much effort if mpv could read timecode track directly from the video file (if available)?

It does, that's what the smpte-timecode property is for. estimated-smpte-timecode will only be available if the video file doesn't have that information

**Comment by Bellavene** [comment]
I believe you misunderstood me.
`video-frame-info/estimated-smpte-timecode` show the timecode starting from 00:00:00:00, but actual video could be started from any time, if a timecode was set in the camera/software. In the example above:
```
Format                                   : QuickTime TC
Time code of first frame                 : 15:52:40:02
Time code of last frame                  : 15:52:53:09
```

**Comment by Bellavene** [comment]
Oh my man, and if somehow we could make `ab-loop-smpte-a` and `ab-loop-smpte-b`, so it would print ab points as smpte values, that would be freaking awesome. Even with my profanity in coding, I could then easily write myself a script for editing EDL lists in mpv and feed them further to editing software, or redirect In/Out points directly to them.

**Comment by Bellavene** [comment]
> > And will it take too much effort if mpv could read timecode track directly from the video file (if available)?
> 
> It does, that's what the smpte-timecode property is for. estimated-smpte-timecode will only be available if the video file doesn't have that information

Or You are stating that it should read the TC track? If so, for some reason `video-frame-info/estimated-smpte-timecode` starts only from 00:00:00:00 for me

**Comment by kasper93** [comment]
> Can we change frame ; character to : ? Would be much a help if you need to copy the timecode from mpv and paste it directly into editing software or EDL list if needed.

`;` vs `:` is standard way to distinguish between non-drop and drop frame timecodes.

> And will it take too much effort if mpv could read timecode track directly from the video file (if available)?

Currently we read timecode as exported from ffmpeg as frame side-data `AV_FRAME_DATA_S12M_TIMECODE`. If it is not available for your file, probably better to ask on ffmpeg side to add it.

Note that it may happen that not every-frame has timecode. This happens with GOP timecodes. Not sure how SMPTE behaves. If you see it for "some" frames, we could sync estimated timecode to it. But only if it is possible and would need to have a file like that.

Long story short, if ffmpeg decodes timecode and send it as side-data mpv will display it.

**Comment by Bellavene** [comment]
1. True, I must use shell scripts then to change it, ok.
2. Sad, but understandable. I will make a request on ffmpeg page then, when I get free time.

**Comment by NicTanghe** [comment]
Hello haven't updated or checked just taking this opertunity to jump in and ask.

Anny possibility to add a dropdown to the ui where we can select showing the current framenumber instead of the timecode ?

**Comment by Bellavene** [comment]
Meta+t no-osd cycle-values osd-msg1 $${estimated-frame-count} ""

**Comment by Bellavene** [comment]
`Meta+t no-osd cycle-values osd-msg1 $${estimated-frame-number} ""`
